File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        max_sum = float("-inf")
        def calculate_max_sum(node):
            nonlocal max_sum
            if not node:
                return 0
            left_sum = max(0,calculate_max_sum(node.left))
            right_sum = max(0,calculate_max_sum(node.right))
            max_sum = max(max_sum,left_sum+right_sum+node.val)
            return max(left_sum,right_sum) + node.val
        calculate_max_sum(root)
        return max_sum

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function maxPathSum(root: TreeNode | null): number {
    if(!root) return 0
    let maxSum = -Infinity;
    function calculateMaxSum(node){
        if(!node) return 0
        let leftBranchSum = Math.max(0,calculateMaxSum(node.left));
        let rightBranchSum = Math.max(0,calculateMaxSum(node.right));
        
        maxSum = Math.max(maxSum,leftBranchSum+node.val+rightBranchSum);
        return Math.max(0,Math.max(leftBranchSum, rightBranchSum)) + node.val
    }
    calculateMaxSum(root);
    return maxSum;
};

